src/CodeReview.Orchestrator/Services/ContainerLogListener.cs

```diff
@@ -0,0 +1,88 @@
+﻿using System;
+using System.Buffers;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Docker.DotNet;
+using Microsoft.Extensions.Logging;
+
+namespace GodelTech.CodeReview.Orchestrator.Services
+{
+    public sealed class ContainerLogListener : IContainerLogListener
+    {
+        private readonly ILogger<ContainerLogListener> _logger;
+
+        private readonly string _containerId;
+        private readonly MultiplexedStream _stream;
+        private readonly CancellationTokenSource _cancellationTokenSource;
+
+        private Task _listenTask;
+
+        public ContainerLogListener(ILogger<ContainerLogListener> logger,
+            long waitTimeoutSeconds,
+            string containerId, 
+            MultiplexedStream stream)
+        {
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            _containerId = containerId ?? throw new ArgumentNullException(nameof(containerId));
+            _stream = stream ?? throw new ArgumentNullException(nameof(stream));
+
+            _cancellationTokenSource = new CancellationTokenSource(TimeSpan.FromSeconds(waitTimeoutSeconds));
+        }
+
+        public void StartListening()
+        {
+            if (_listenTask != null)
+                throw new InvalidOperationException("Already listening a stream");
+
+            _listenTask = Task.Factory.StartNew(Listen, _cancellationTokenSource.Token, TaskCreationOptions.LongRunning);
+        }
+
+        public void StopListening()
+        {
+            if (_listenTask == null || _cancellationTokenSource.IsCancellationRequested)
+                return;
+
+            _cancellationTokenSource.Cancel();
```

Comment:
I'm wondering why don't we want to `_listenTask` here? It my mind should have something similar to this here 
`_listenTask.GetAwaiter().GetResult()` (e.g. to get inner exception rather than AggregateException).

Reply:
This should run as a background worker. Like a demon thread. I don't think we need to wait for the operation to complete. We just need to tell the thread that it should be stopped.